doc/developer/design/20230714_optimizer_interface.md

```diff
@@ -0,0 +1,336 @@
+  output. This allows to have more than one entrypoints to a pipeline.
+- The concrete types used for stage results are opaque structs.
+  - The only way to construct a struct is by running the `Optimize` stage that
+    produces it. This ensures that client code cannot interfere with the
+    pipeline.
+  - In general, the internals of these structs can be accessed only behind a
+    shared reference. This ensures that client code can look up information from
+    intermediate stages but in general cannot modify it.
+  - Timestamp selection is modeled as a conversion between structs that are
+    adjacent in the pipeline.
+  - The internals of the struct representing the result of the final stage of
+    the optimization pipeline for a specific type can be destructed.
+- The `Send + Sync` trait bounds on the `Self` and `From` types ensure that
+  `Optimize` instances can be passed to different threads.
+
+As an example, here is the definition of the `OptimizeMaterializedView` struct
+which implements the optimizer pipeline for `CREATE MATERIALIZED VIEW`
+statements:
+
+```rust
+pub struct OptimizeMaterializedView {
+    /// A typechecking context to use throughout the optimzier pipeline.
+    typecheck_ctx: TypecheckContext,
+    /// A snapshot of the catalog state.
+    catalog: Arc<Catalog>,
+    /// A snapshot of the compute instance that will run the dataflows.
+    compute_instance: ComputeInstanceSnapshot,
+    /// A durable GlobalId to be used with the exported materialized view sink.
+    exported_sink_id: GlobalId,
+    /// A transient GlobalId to be used when constructing the dataflow.
+    internal_view_id: GlobalId,
+    /// The resulting column names.
+    column_names: Vec<ColumnName>,
+    /// A human-readable exposed internally (useful for debugging).
+    debug_name: String,
+    // Optimizer flags.
+    flags: OptimizerFlags,
+}
+```
+
+The following diagram illustrates the pipeline implemented by `OptimizeMaterializedView`
+
+```mermaid
+flowchart TD
+
+HRE(HirRelationExpr) --> LMP(LocalMirPlan)
+LMP --> GM1("GlobalMirPlan< Unresolved >")
+OMRE(OptimizedMirRelationExpr) --> GM1
+GM1 -.-> GM2("GlobalMirPlan< Resolved >")
+GM2 --> GL(GlobalLirPlan)
+```
+
+Solid edges represent transitions done by `optimize` calls on the various
+`Optimize<From>` implementations on `OptimizeMaterializedView`. The dotted edge
+represents a transition between a non-timestamped `GlobalMirPlan` to a
+timestamped `GlobalMirPlan` that the client can perform after determining the
+execution timestamp for the given statement.
+
+The `LocalMirPlan`, `GlobalMirPlan`, and `GlobalLirPlan` types encapsulate
+intermediate results that wrap the resulting plan and strucutre and associated
+metadata.
+
+The pipeline can be entered both from an `HirRelationExpr` and from an
+`OptimizedMirRelationExpr` — the latter is needed because the
+`Coordinator::bootstrap` code path starts from a catalog item that stores an
+already locally optimized `MirRelationExpr`.
```

Comment:
Thanks for providing the example! Could you also stub out the `Optimize` impls for `OptimizeMaterializedView`?

I assume it would look something like
```rust
impl Optimize<HirRelationExpr> for OptimizeMaterializedView {
    To = GlobalMirPlan<Unresolved>,
    ...
}
impl Optimize<OptimizedMirRelationExpr> for OptimizeMaterializedView {
    To = GlobalMirPlan<Unresolved>,
    ...
}
impl Optimize<GlobalMirPlan<Resolved>> for OptimizeMaterializedView {
    To = GlobalLirPlan,
    ...
}
```